keras_nlp/src/models/phi3/phi3_rotary_embedding.py

```diff
@@ -0,0 +1,149 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+
+from keras_nlp.src.backend import ops
+from keras_nlp.src.layers.modeling.rotary_embedding import RotaryEmbedding
+
+
+class Phi3SuScaledRotaryEmbedding(RotaryEmbedding):
+    """SuRotary positional encoding layer.
+
+    Args:
+        max_sequence_length: int. The maximum sequence length that this
+            model might ever be used with.
+        original_max_sequence_length: int. The maximum sequence length that
+            this model was trained with.
+        inverese_freq_short_factor List[float]: List of factors used to adjust
+            rope frequencies when the `rope_scaling_type` is `"su"`. List must
+            be of length `hidden_dim//num_query_heads//2`. It is used when
+            `sequence_length` is smaller than `original_max_sequence_length`.
+            Defaults to `None`.
+        inverese_freq_long_factor List[float]: List of factors used to adjust
+            rope frequencies when the `rope_scaling_type` is `"su"`. List must
+            be of length `hidden_dim//num_query_heads//2`. It is used when
+            `sequence_length` is larger than `original_max_sequence_length`.
+            Defaults to `None`.
+        max_wavelength: int. The maximum angular wavelength of the sine/cosine
+            curves.
+
+    Call arguments:
+        inputs: The tensor inputs to apply the embedding to. This can have
+            any shape, but must contain both a sequence and feature axis. The
+            rotary embedding will be applied to `inputs` and returned.
+        start_index: An integer or integer tensor. The starting position to
+            compute the rotary embedding from. This is useful during cached
+            decoding, where each position is predicted separately in a loop.
+
+    References:
+     - [Phi-3-mini-128k-instruct original implementation](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/0693e0b867d29e7318280ddd3ff9d5e66698f488/modeling_phi3.py#L142)
+    """
+
+    def __init__(
+        self,
+        max_sequence_length=4096,
+        original_max_sequence_length=4096,
+        inverese_freq_short_factor=None,
+        inverese_freq_long_factor=None,
+        max_wavelength=10000,
+        **kwargs
+    ):
+        super().__init__(max_wavelength=max_wavelength, **kwargs)
+        self.max_sequence_length = max_sequence_length
+        self.original_max_sequence_length = original_max_sequence_length
+
+        scaling_factor = (
+            self.max_sequence_length / self.original_max_sequence_length
+        )
+        if scaling_factor <= 1.0:
+            self.embedding_scaling_factor = 1.0
+        else:
+            self.embedding_scaling_factor = math.sqrt(
+                1
+                + math.log(scaling_factor)
+                / math.log(self.original_max_sequence_length)
+            )
+
+        if inverese_freq_short_factor is not None:
```

Comment:
Why can't we just store the simply pythonic thing here (either none or a list of strings) and convert to tensor as needed down in the call parts?

That might look a little cleaner in code, and would be much better for saving the config of this layer. No one will be directly saving with this layer today, but if they did, it would raise because a tensor cannot be converted to json.